### DISCUSSION TITLE
How to model start and end locations for vehicles that are not depot and not any of pickup or drop-off locations

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#3882] [Routing (and legacy CP) questions]
Hello fellow VRP-ers!

I haver a VRP issue where on some days vehicles start in arbitrary locations on the map, which are NOT depot and are NOT any pickup and drop-off locations (think of it as the driver kept the van at home and starts a new day from home). How do I model that? I am not sure where to start modeling the distance from driver's "home" (or last known location)? I need VRP to efficiently find first route for the vehicles given their initial location.

Thank you.

**Comment by odashevsky** [comment]
Anybody? Any ideas are appreciated.

**Comment by jmarca** [comment]
If you know the start locations, then just set those as the depot for
that vehicle.

Am I missing something?


On Tue, Aug 08, 2023 at 02:26:57PM -0700, Oleg Dashevsky wrote:
> Anybody? Any ideas are appreciated.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3882#discussioncomment-6673689
> You are receiving this because you are subscribed to this thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC

**Reply by odashevsky** [reply]
Hi,
Well, how do I do that? I may end up with 100 different depots (if I have 100 vehicles). Each of the depots is not any of the pickup or dropoff locations. Not even sure how to build RoutingIndexManager, as it takes either "int depot" or "int[] starts, int[] ends", which are nodes on DistanceMatrix, which are not in my case points that need to be "visited" by the vehicles.

**Comment by jmarca** [comment]
Yes that is exactly it. The int starts and ends are the locations of your vehicles start and end points. Try it small scale first, then ramp up. And for SURE you need to make sure the caching is turned on, or the solver will take forever with that many vehicles. JamesOn Aug 8, 2023, at 14:40, Oleg Dashevsky ***@***.***> wrote:﻿
Hi,
Well, how do I do that? I may end up with 100 different depots (if I have 100 vehicles). Each of the depots is not any of the pickup or dropoff locations. Not even sure how to build RoutingIndexManager, as it takes either "int depot" or "int[] starts, int[] ends", which are nodes on DistanceMatrix, which are not in my case points that need to be "visited" by the vehicles.

**Reply by odashevsky** [reply]
Well, here's the thing and please tell me where I am wrong. 
"int[] starts, int[] ends" (of RoutingIndexManager constructor)  are arrays having the lengths of how many vehicles I have. In this case 100 (just an example). Each entry in each array refers to a location's **node number** in Distance Matrix, which is LxL matrix, where L is the number of nodes where a pickup or dropoff has to take place. Well, none of the starting or ending locations of the vehicle is ANY of these nodes that need to be "visited". So, I am not sure what to populate "int[] starts" and "int[] ends" with. Am I making sense? 
And as always, thanks for your help!

**Comment by jmarca** [comment]
Yes and no. Say you have L nodes to visit, and M starting points. It is entirely up to you how you organize them, but I like to put L before M. Then the first LxL entries typically correspond to the nodes to be visited. Then the remaining blocks (L to M, M to L, and M to M) represent movement to and from the depot nodes. (I make the M to M all infinity, just to be safe.)Try it with a small problem and see what is happening. You can put little print statements in the callback to see what the solver asks for (index to index) and what the call to manager.IndexToNode(index) produces. JamesOn Aug 9, 2023, at 07:13, Oleg Dashevsky ***@***.***> wrote:﻿
Well, here's the thing and please tell me where I am wrong.
"int[] starts, int[] ends" (of RoutingIndexManager constructor)  are arrays having the lengths of how many vehicles I have. In this case 100 (just an example). Each entry in each array refers to a location's node number in Distance Matrix, which is LxL matrix, where L is the number of nodes where a pickup or dropoff has to take place. Well, none of the starting or ending locations of the vehicle is ANY of these nodes that need to be "visited". So, I am not sure what to populate "int[] starts" and "int[] ends" with. Am I making sense?
And as always, thanks for your help!
